File: Algorithm/077. Combinations.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-17 10:03:19
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-17 10:03:44
# @Email: liangchaowu5@gmail.com

# backtracking
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        result = []
        tmp = []
        self.helper(n+1,k,1,result,tmp)
        return result
        
    def helper(self,n,k,begin,result,tmp):
        if len(tmp) == k:
            result.append(tmp[:])
        for i in xrange(begin,n):
            tmp.append(i)
            self.helper(n,k,i+1,result,tmp)
            tmp.remove(i)
        

File: Algorithm/Java/077. Combinations.java
/**
* Author: WuLC
* Date:   2017-01-22 21:19:50
* Last modified by:   WuLC
* Last Modified time: 2017-01-22 21:20:08
* Email: liangchaowu5@gmail.com
*/

// dfs
public class Solution 
{
    private List<List<Integer>> result;
    private List<Integer> tmp;
    public List<List<Integer>> combine(int n, int k) 
    {
        result = new ArrayList<List<Integer>>();
        tmp = new ArrayList<Integer>();
        dfs(1, n, k);
        return result;
    }
    
    private void dfs(int curr, int n, int k)
    {
        if (tmp.size() == k)
        {
            result.add(new ArrayList(tmp));
            return;
        }
        for(int i=curr; i<=n; i++)
        {
            tmp.add(i);
            dfs(i+1, n, k);
            tmp.remove(tmp.size()-1);
        }
    }
}